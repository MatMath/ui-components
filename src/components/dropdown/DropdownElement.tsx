import * as React from 'react';
import styled from 'styled-components';
import { StyledText } from '@ions/text/StyledText';
import { dashGreen00, midGreen05, white } from '@colors';

interface Props {
  onClick: (e: React.MouseEvent) => void;
  elementClassName?: string;
}

const Wrapper = styled.button`
  color: ${dashGreen00};
  line-height: 20px;
  font-size: 16px;
  cursor: pointer;
  border: transparent 0px;
  outline: none;
  display: flex;
  box-sizing: border-box;
  height: 40px;
  flex-shrink: 0;
  padding: 8px 16px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  background-color: ${white};

  &:hover {
    background-color: ${midGreen05};
  }
`;

export const DropdownElement: React.FC<Props> = props => {
  return (
    <Wrapper
      type='button'
      onClick={props.onClick}
      className={props.elementClassName}
    >
      <StyledText>{props.children}</StyledText>
    </Wrapper>
  );
};
