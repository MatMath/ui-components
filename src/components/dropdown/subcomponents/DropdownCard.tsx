import styled, { css } from 'styled-components';
import { dashGreen04, white } from '@colors';

interface Props {
  maxHeight?: number;
  className?: string;
}

export const DropdownCard = styled.div<Props>`
  display: flex;
  flex-direction: column;
  background-color: ${white};
  border: 1px solid ${dashGreen04};
  box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.2);
  border-radius: 4px;
  padding: 8px 0;
  ${props =>
    props.maxHeight &&
    css`
      max-height: ${props.maxHeight}px;
      overflow-y: scroll;
    `}
`;
