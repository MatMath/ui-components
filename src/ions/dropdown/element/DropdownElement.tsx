import * as React from 'react';
import { StyledText } from '@ions/text/StyledText';
import styled from 'styled-components';
import { dashGreen00, dashGreen06 } from '@colors';

interface Props {
  onClick: (e: React.MouseEvent) => void;
  elementClassName?: string;
}

const Wrapper = styled.button`
  color: ${dashGreen00};
  line-height: 20px;
  font-size: 16px;
  cursor: pointer;
  border: transparent 0px;
  outline: none;
  display: flex;
  box-sizing: border-box;
  height: 36px;
  flex-shrink: 0;
  padding: 8px 16px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;

  &:hover {
    background-color: ${dashGreen06};
  }
`;

const DropdownElement: React.FC<Props> = props => {
  return (
    <Wrapper
      type='button'
      onClick={props.onClick}
      className={props.elementClassName}
    >
      <StyledText>{props.children}</StyledText>
    </Wrapper>
  );
};

export default DropdownElement;
