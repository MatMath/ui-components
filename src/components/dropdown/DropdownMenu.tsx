import * as React from 'react';
import styled from 'styled-components';
import { DropdownCard } from './subcomponents/DropdownCard';
import { DockingSide, Placement } from './types';
import {
  getControllerAndMenuDimensions,
  getDropdownPosition,
  isDropdownOutOfContainer
} from './helpers';
import {
  getContainerBoundaries,
  getElementPlacement
} from '@utility/positionCompute';

export interface DropdownMenuProps {
  isOpen: boolean;
  controllerRef: React.RefObject<HTMLElement>;
  /**
   * using menuMaxHeight will add max-height and overflow-y: scroll css properties
   */
  menuMaxHeight?: number;
  dockingSide?: DockingSide;
  placement?: Placement;
  containerRef?: React.RefObject<HTMLElement>;
  menuClassName?: string;
}

export interface MenuWrapperProps {
  placement: Placement;
  dockingSide: DockingSide;
  isReadyForDisplay: boolean;
}

const MenuWrapper = styled.div<MenuWrapperProps>`
  position: absolute;
  opacity: ${props => (props.isReadyForDisplay ? 1 : 0)};
  ${getDropdownPosition}
  max-width: 318px;
`;

const DEFAULT_PLACEMENT = 'bottom';
const POSSIBLE_PLACEMENTS: Placement[] = ['top', 'bottom'];

export const DropdownMenu = ({
  isOpen,
  controllerRef,
  menuMaxHeight,
  dockingSide,
  placement: dropdownPlacement,
  containerRef,
  menuClassName,
  children
}: React.PropsWithChildren<DropdownMenuProps>) => {
  const [placement, setPlacement] = React.useState<Placement | null>(null);
  const menuRef = React.useRef<HTMLDivElement>(null);
  const prevIsOpen = React.useRef(false);

  React.useEffect(() => {
    if (isOpen === prevIsOpen.current) {
      return;
    }
    prevIsOpen.current = isOpen;
    setPlacement(null);

    if (!menuRef.current || !controllerRef.current) {
      return;
    }
    const elementDimensions = getControllerAndMenuDimensions({
      menuRef,
      controllerRef
    });

    const containerElement = (containerRef && containerRef.current) || window;
    const containerDimensions = getContainerBoundaries(containerElement);

    try {
      const newPlacement = getElementPlacement({
        isElementOutOfContainerMethod: isDropdownOutOfContainer,
        placement: dropdownPlacement || DEFAULT_PLACEMENT,
        possibilities: POSSIBLE_PLACEMENTS,
        elementDimensions,
        containerDimensions
      });
      setPlacement(newPlacement);
    } catch (e) {
      console.error('Error while computing DropdownMenu position', e);
      setPlacement(DEFAULT_PLACEMENT);
    }
  }, [containerRef, controllerRef, dropdownPlacement, isOpen]);

  return (
    <>
      {isOpen && (
        <MenuWrapper
          placement={placement || DEFAULT_PLACEMENT}
          ref={menuRef}
          dockingSide={dockingSide || 'left'}
          isReadyForDisplay={placement !== null}
        >
          <DropdownCard maxHeight={menuMaxHeight} className={menuClassName}>
            {children}
          </DropdownCard>
        </MenuWrapper>
      )}
    </>
  );
};
