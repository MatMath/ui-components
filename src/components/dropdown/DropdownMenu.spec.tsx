import * as React from 'react';
import { mount } from 'enzyme';
import { DropdownCard } from './subcomponents/DropdownCard';
import { DropdownMenu, DropdownMenuProps } from './DropdownMenu';
import { DropdownElement } from './DropdownElement';

describe('<DropdownMenu>', () => {
  const getDropdownMenuWrapper = (props: DropdownMenuProps) => {
    const controllerRef = props.controllerRef as React.RefObject<
      HTMLDivElement
    >;
    return mount(
      <div>
        <div ref={controllerRef}>Dropdown Menu controller</div>
        <DropdownMenu
          controllerRef={controllerRef}
          isOpen={props.isOpen}
          menuMaxHeight={props.menuMaxHeight}
        >
          <DropdownElement onClick={() => {}}>Element 1</DropdownElement>
          <DropdownElement onClick={() => {}}>Element 2</DropdownElement>
        </DropdownMenu>
      </div>
    );
  };
  describe('Global render', () => {
    const controllerRef = React.createRef<HTMLDivElement>();
    it('should have scroll enabled when using max height', () => {
      const menuMaxHeight = 100;
      const dropdownMenuWrapper = getDropdownMenuWrapper({
        isOpen: true,
        menuMaxHeight: menuMaxHeight,
        controllerRef: controllerRef
      });
      expect(
        window
          .getComputedStyle(dropdownMenuWrapper.find(DropdownCard).getDOMNode())
          .getPropertyValue('overflow-y')
      ).toEqual('scroll');

      expect(dropdownMenuWrapper).toMatchSnapshot();
    });

    it('should not have scroll enabled when not using max height', () => {
      const dropdownMenuWrapper = getDropdownMenuWrapper({
        isOpen: true,
        controllerRef: controllerRef
      });
      expect(
        window
          .getComputedStyle(dropdownMenuWrapper.find(DropdownCard).getDOMNode())
          .getPropertyValue('overflow-y')
      ).not.toEqual('scroll');

      expect(dropdownMenuWrapper).toMatchSnapshot();
    });
  });
});
