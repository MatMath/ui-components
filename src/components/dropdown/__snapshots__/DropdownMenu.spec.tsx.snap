// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownMenu> Global render should have scroll enabled when using max height 1`] = `
<div>
  <div>
    Dropdown Menu controller
  </div>
  <Component
    controllerRef={
      Object {
        "current": <div>
          Dropdown Menu controller
        </div>,
      }
    }
    isOpen={true}
    menuMaxHeight={100}
  >
    <styled.div
      dockingSide="left"
      isReadyForDisplay={true}
      placement="bottom"
    >
      <StyledComponent
        dockingSide="left"
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bwzfXH",
              "isStatic": false,
              "lastClassName": "izZuan",
              "rules": Array [
                "
  position: absolute;
  opacity: ",
                [Function],
                ";
  ",
                [Function],
                "
  max-width: 318px;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bwzfXH",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={
          Object {
            "current": <div
              class="sc-bwzfXH izZuan"
            >
              <div
                class="sc-bdVaJa jOpScq"
              >
                <button
                  class="sc-bxivhb kTePkg"
                  type="button"
                >
                  <span
                    class="sc-htpNat cAleLB"
                  >
                    Element 1
                  </span>
                </button>
                <button
                  class="sc-bxivhb kTePkg"
                  type="button"
                >
                  <span
                    class="sc-htpNat cAleLB"
                  >
                    Element 2
                  </span>
                </button>
              </div>
            </div>,
          }
        }
        isReadyForDisplay={true}
        placement="bottom"
      >
        <div
          className="sc-bwzfXH izZuan"
        >
          <styled.div
            maxHeight={100}
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bdVaJa",
                    "isStatic": false,
                    "lastClassName": "jOpScq",
                    "rules": Array [
                      "
  display: flex;
  flex-direction: column;
  background-color: ",
                      "#FFFFFF",
                      ";
  border: 1px solid ",
                      "#CFD7D8",
                      ";
  box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.2);
  border-radius: 4px;
  padding: 8px 0;
  ",
                      [Function],
                      "
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bdVaJa",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              maxHeight={100}
            >
              <div
                className="sc-bdVaJa jOpScq"
              >
                <Component
                  onClick={[Function]}
                >
                  <styled.button
                    onClick={[Function]}
                    type="button"
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bxivhb",
                            "isStatic": true,
                            "lastClassName": "kTePkg",
                            "rules": Array [
                              "
  color: ",
                              "#0E353D",
                              ";
  line-height: 20px;
  font-size: 16px;
  cursor: pointer;
  border: transparent 0px;
  outline: none;
  display: flex;
  box-sizing: border-box;
  height: 40px;
  flex-shrink: 0;
  padding: 8px 16px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  background-color: ",
                              "#FFFFFF",
                              ";

  &:hover {
    background-color: ",
                              "#D9E6E9",
                              ";
  }
",
                            ],
                          },
                          "displayName": "styled.button",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bxivhb",
                          "target": "button",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="sc-bxivhb kTePkg"
                        onClick={[Function]}
                        type="button"
                      >
                        <styled.span>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-htpNat",
                                  "isStatic": true,
                                  "lastClassName": "cAleLB",
                                  "rules": Array [
                                    "
  font-family: inherit;
  font-weight: 600;
",
                                  ],
                                },
                                "displayName": "styled.span",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-htpNat",
                                "target": "span",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <span
                              className="sc-htpNat cAleLB"
                            >
                              Element 1
                            </span>
                          </StyledComponent>
                        </styled.span>
                      </button>
                    </StyledComponent>
                  </styled.button>
                </Component>
                <Component
                  onClick={[Function]}
                >
                  <styled.button
                    onClick={[Function]}
                    type="button"
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bxivhb",
                            "isStatic": true,
                            "lastClassName": "kTePkg",
                            "rules": Array [
                              "
  color: ",
                              "#0E353D",
                              ";
  line-height: 20px;
  font-size: 16px;
  cursor: pointer;
  border: transparent 0px;
  outline: none;
  display: flex;
  box-sizing: border-box;
  height: 40px;
  flex-shrink: 0;
  padding: 8px 16px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  background-color: ",
                              "#FFFFFF",
                              ";

  &:hover {
    background-color: ",
                              "#D9E6E9",
                              ";
  }
",
                            ],
                          },
                          "displayName": "styled.button",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bxivhb",
                          "target": "button",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="sc-bxivhb kTePkg"
                        onClick={[Function]}
                        type="button"
                      >
                        <styled.span>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-htpNat",
                                  "isStatic": true,
                                  "lastClassName": "cAleLB",
                                  "rules": Array [
                                    "
  font-family: inherit;
  font-weight: 600;
",
                                  ],
                                },
                                "displayName": "styled.span",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-htpNat",
                                "target": "span",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <span
                              className="sc-htpNat cAleLB"
                            >
                              Element 2
                            </span>
                          </StyledComponent>
                        </styled.span>
                      </button>
                    </StyledComponent>
                  </styled.button>
                </Component>
              </div>
            </StyledComponent>
          </styled.div>
        </div>
      </StyledComponent>
    </styled.div>
  </Component>
</div>
`;

exports[`<DropdownMenu> Global render should not have scroll enabled when not using max height 1`] = `
<div>
  <div>
    Dropdown Menu controller
  </div>
  <Component
    controllerRef={
      Object {
        "current": <div>
          Dropdown Menu controller
        </div>,
      }
    }
    isOpen={true}
  >
    <styled.div
      dockingSide="left"
      isReadyForDisplay={true}
      placement="bottom"
    >
      <StyledComponent
        dockingSide="left"
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bwzfXH",
              "isStatic": false,
              "lastClassName": "izZuan",
              "rules": Array [
                "
  position: absolute;
  opacity: ",
                [Function],
                ";
  ",
                [Function],
                "
  max-width: 318px;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bwzfXH",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={
          Object {
            "current": <div
              class="sc-bwzfXH izZuan"
            >
              <div
                class="sc-bdVaJa bIoTfe"
              >
                <button
                  class="sc-bxivhb kTePkg"
                  type="button"
                >
                  <span
                    class="sc-htpNat cAleLB"
                  >
                    Element 1
                  </span>
                </button>
                <button
                  class="sc-bxivhb kTePkg"
                  type="button"
                >
                  <span
                    class="sc-htpNat cAleLB"
                  >
                    Element 2
                  </span>
                </button>
              </div>
            </div>,
          }
        }
        isReadyForDisplay={true}
        placement="bottom"
      >
        <div
          className="sc-bwzfXH izZuan"
        >
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bdVaJa",
                    "isStatic": false,
                    "lastClassName": "bIoTfe",
                    "rules": Array [
                      "
  display: flex;
  flex-direction: column;
  background-color: ",
                      "#FFFFFF",
                      ";
  border: 1px solid ",
                      "#CFD7D8",
                      ";
  box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.2);
  border-radius: 4px;
  padding: 8px 0;
  ",
                      [Function],
                      "
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bdVaJa",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="sc-bdVaJa bIoTfe"
              >
                <Component
                  onClick={[Function]}
                >
                  <styled.button
                    onClick={[Function]}
                    type="button"
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bxivhb",
                            "isStatic": true,
                            "lastClassName": "kTePkg",
                            "rules": Array [
                              "
  color: ",
                              "#0E353D",
                              ";
  line-height: 20px;
  font-size: 16px;
  cursor: pointer;
  border: transparent 0px;
  outline: none;
  display: flex;
  box-sizing: border-box;
  height: 40px;
  flex-shrink: 0;
  padding: 8px 16px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  background-color: ",
                              "#FFFFFF",
                              ";

  &:hover {
    background-color: ",
                              "#D9E6E9",
                              ";
  }
",
                            ],
                          },
                          "displayName": "styled.button",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bxivhb",
                          "target": "button",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="sc-bxivhb kTePkg"
                        onClick={[Function]}
                        type="button"
                      >
                        <styled.span>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-htpNat",
                                  "isStatic": true,
                                  "lastClassName": "cAleLB",
                                  "rules": Array [
                                    "
  font-family: inherit;
  font-weight: 600;
",
                                  ],
                                },
                                "displayName": "styled.span",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-htpNat",
                                "target": "span",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <span
                              className="sc-htpNat cAleLB"
                            >
                              Element 1
                            </span>
                          </StyledComponent>
                        </styled.span>
                      </button>
                    </StyledComponent>
                  </styled.button>
                </Component>
                <Component
                  onClick={[Function]}
                >
                  <styled.button
                    onClick={[Function]}
                    type="button"
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bxivhb",
                            "isStatic": true,
                            "lastClassName": "kTePkg",
                            "rules": Array [
                              "
  color: ",
                              "#0E353D",
                              ";
  line-height: 20px;
  font-size: 16px;
  cursor: pointer;
  border: transparent 0px;
  outline: none;
  display: flex;
  box-sizing: border-box;
  height: 40px;
  flex-shrink: 0;
  padding: 8px 16px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  background-color: ",
                              "#FFFFFF",
                              ";

  &:hover {
    background-color: ",
                              "#D9E6E9",
                              ";
  }
",
                            ],
                          },
                          "displayName": "styled.button",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bxivhb",
                          "target": "button",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="sc-bxivhb kTePkg"
                        onClick={[Function]}
                        type="button"
                      >
                        <styled.span>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-htpNat",
                                  "isStatic": true,
                                  "lastClassName": "cAleLB",
                                  "rules": Array [
                                    "
  font-family: inherit;
  font-weight: 600;
",
                                  ],
                                },
                                "displayName": "styled.span",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-htpNat",
                                "target": "span",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <span
                              className="sc-htpNat cAleLB"
                            >
                              Element 2
                            </span>
                          </StyledComponent>
                        </styled.span>
                      </button>
                    </StyledComponent>
                  </styled.button>
                </Component>
              </div>
            </StyledComponent>
          </styled.div>
        </div>
      </StyledComponent>
    </styled.div>
  </Component>
</div>
`;
